### ISSUE TITLE
Potential bug or code clarity?

### ISSUE THREAD

**Issue by MortenHolmRep** [graphnet-team/graphnet#417]
**Describe the bug**
Compared to the Zenith task, the Azimuth task seem to have an inversed inheritance?

**Code in current form**
```
class AzimuthReconstruction(AzimuthReconstructionWithKappa):
    """Reconstructs azimuthal angle."""

    # Requires two features: untransformed points in (x,y)-space.
    nb_inputs = 2

    def _forward(self, x: Tensor) -> Tensor:
        # Transform outputs to angle and prepare prediction
        res = super()._forward(x)
        angle = res[:, 0].unsqueeze(1)
        kappa = res[:, 1]
        sigma = torch.sqrt(1.0 / kappa)
        beta = 1e-3
        kl_loss = torch.mean(sigma**2 - torch.log(sigma) - 1)
        self._regularisation_loss += beta * kl_loss
        return angle
        
class AzimuthReconstructionWithKappa(Task):
    """Reconstructs azimuthal angle and associated kappa (1/var)."""

    # Requires two features: untransformed points in (x,y)-space.
    nb_inputs = 2

    def _forward(self, x: Tensor) -> Tensor:
        # Transform outputs to angle and prepare prediction
        kappa = torch.linalg.vector_norm(x, dim=1) + eps_like(x)
        angle = torch.atan2(x[:, 1], x[:, 0])
        angle = torch.where(
            angle < 0, angle + 2 * np.pi, angle
        )  # atan(y,x) -> [-pi, pi]
        return torch.stack((angle, kappa), dim=1)
```

Dependency seem reversed

**Code as I would imagine it should be**
```
class AzimuthReconstruction(Task):
    """Reconstructs azimuthal angle."""

    # Requires two features: untransformed points in (x,y)-space.
    nb_inputs = 2

    def _forward(self, x: Tensor) -> Tensor:
        angle = torch.atan2(x[:, 1], x[:, 0])
        angle = torch.where(
            angle < 0, angle + 2 * np.pi, angle
        )  # atan(y,x) -> [-pi, pi]
        return angle



class AzimuthReconstructionWithKappa(AzimuthReconstruction):
    """Reconstructs azimuthal angle and associated kappa (1/var)."""

    # Requires two features: untransformed points in (x,y)-space.
    nb_inputs = 2

    def _forward(self, x: Tensor) -> Tensor:
        # Transform outputs to angle and prepare prediction
        angle = super()._forward(x)
        kappa = torch.linalg.vector_norm(x, dim=1) + eps_like(x)
        sigma = torch.sqrt(1.0 / kappa)
        beta = 1e-3
        kl_loss = torch.mean(sigma**2 - torch.log(sigma) - 1)
        self._regularisation_loss += beta * kl_loss
        return torch.stack((angle, kappa), dim=1)
```

**Comment by asogaard** [comment]
Hi @MortenHolmRep,

Thanks for the keen eye. In both case, the idea behind each inheritance is to start from the simpler/more natural case than then specialise/inherit to the less natural one. The difference between the two cases is that:

* Azimuth prediction is most naturally done in 2D due to the circular boundary conditions. If you're predicting an arbitrary vector in the 2D-plane, this is already the exact form you want for your output for the von Mises-Fisher loss: the direction of the vector in the plane gives the angle; the magnitude of the vector is equal to kappa. If you don't care about uncertainty (kappa), then you remove one "degree of freedom" so to speak, and you want to constrain the vector to the unit circle and just use the direction in your reconstruction task. That's what's being done in the inheriting class.
* Zenith prediction is comparatively simple, as there is no circular boundary conditions: you just spit out a 1D number in [0, π]. Because you don't gain anything from starting from the more involved situation with uncertainty in this case, we don't. The inheriting class then adds an independent kappa variable, and that's it.

So, while it would be intuitive to flip the order of inheritance for azimuth prediction, I think would would have to do something a little hacky to constrain the prediction to the unit circle in the first place, and then add that degree of freedom back in later, which seems at little circular — at least to me. But that is not to say that there can't be other ways of going about it that might make the structure of the two classes a bit more similar. 

Does that make sens, @MortenHolmRep? :)
